### ISSUE TITLE
Modify project completion metric to use classification counts

### ISSUE THREAD

**Issue by camallen** [zooniverse/Panoptes#2492]
As per the comment from PFE - https://github.com/zooniverse/Panoptes-Front-End/issues/4093#issuecomment-336896345

Perhaps we should summate the workflow metrics according to the prefs now set instead of using the default retired ratio.

**Comment by mrniaboc** [comment]
I'm absolutely behind this 100%. If I was the kind of person who wrote in all caps, I would be right now.

**Comment by camallen** [comment]
The current calculation doesn't take into account the proportion of each subjects subject counts, instead we just average out the the sum of the active workflow percentages. https://github.com/zooniverse/Panoptes/blob/fddc6e38dd4d1d61f3b767b2b3904d5496286a8a/app/workers/calculate_project_completeness_worker.rb#L42

If i have 2 active workflows:
1. workflow 1 = 1 retired / 5 subjects = 0.2
0. workflow 2 = 9 retired / 10 subjects = 0.9

The current calc is (0.2 + 0.9) / 2 = 0.55, it is actually 10/15 = 0.66. This needs to change to either just the classification counts or we weight the percentages coming from each workflow, e.g. 0.2 of 5/15 and 0.9 of 10/15.

**Comment by camallen** [comment]
Also noting that classification counts may be impacted by early retirement from caesar as well, though it's probably more accurate than what we have.

**Comment by mrniaboc** [comment]
The current set-up continues to confuse researchers and volunteers alike. Got this message from Jules earlier today:

"Just a quick query. The stars on the Solar Stormwatch II homepage has always remained at 0%. It's currently at 37% if you click on the link but not everyone does and a demoralising 0% is all they see. Is this fixable? If not could it be removed just leaving the link? I've had a look to see if the SSW team could do anything but it doesn't look like it."

We have a progress bar by classification count option in the stats page, we really should at lest allow the option for the homepage!

**Comment by camallen** [comment]
Just noting that classifications count can also be problematic if the team changes the retirement limit hald way through a data processing run, https://www.zooniverse.org/talk/17/864027?comment=1426076

**Comment by mrniaboc** [comment]
True, but not as problematic as everyone wondering why projects show 0%
complete all the time even though they have thousands of classifications.

**Comment by mrniaboc** [comment]
This still continues to confuse researchers. Today it was Ed Hawkins, the PI of the Weather Rescue projects. He has released three projects and knows his stuff, but still couldn't figure out why his progress bar was not increasing.

**Comment by mrniaboc** [comment]
Another volunteer complaining that this was making them lose heart on Muon Hunter today. Would really love to make this change at some point soon.

The fact we can show a progress bar based on classifications on our stats page but not the homepage is so frustrating!

**Comment by camallen** [comment]
Not saying we shouldn't modify the completeness metric API side but whatever we change too will have trade-offs as well, especially with external retirement systems (e.g. caesar).

Interestingly the fact this information exists API side for the stats page to work means this change can happen in the UI as well.

**Comment by mrniaboc** [comment]
Yeah, I totally accept it won't makes sense for projects with other retirement rules, however at the moment it makes no sense for the majority of projects.

**Comment by srallen** [comment]
What change can happen UI side? Could you open an issue so we at least don't replicate it in the rewrite?

**Comment by camallen** [comment]
> The fact we can show a progress bar based on classifications on our stats page but not the homepage

**Comment by camallen** [comment]
Apologies, that sent early. The issue is the API project resource has a non-optimal calculation for the completion metric whereas the stats pages calculate workflow completeness differently (configurable) based on expected classification counts against the retirement count. 

The UI could use that information to calculate different progress metrics for presentation to the user.

Perhaps the API project resource should have both metrics as both have trade offs:
1. Subject retirement is a lagging indicator - it confuses folks as it only starts changing after a long time
2. Classification counts may never be reached if caesar is retiring data early vs the exepcted counts and projects will never finish.

**Comment by PmasonFF** [comment]
One option might be to track the classifications per subject for **retired** subjects - This average, times the number of active subjects is a measure of the total classifications needed to retire all subjects regardless of the manner (Caesar, Bayesian pipeline, custom rules) retirement is determined.  EDC = (total expected classifications required - classifications done)/average daily classifications.
I will look at trying to get a average classifications to retirement from looking at the Subjectworkflowstatus. This will be slow but it is not necessary to keep a frequently updated value since it is only used as an average. Perhaps there is some means to determine if there is a major change to the workflows or retirement means that would affect the number, otherwise run with a weekly or even less frequently updated value.
Even better would be a direct calculation of the classifications required to complete unretired subjects, based on the average classifications to retirement, and the current number of classifications for each subject, but this number would have to be calculated every time an EDC was to be updated.

These EDC values also need to be cleaned of the increasing number of "training/practice " workflows with very high retirement limits and never any completeness  eg Earthquake detective runs 3 practice  workflows and rotates through small active subject sets of real data - the ECD numbers for the active subjects are be skewed because the the practice workflow subjects are lumped in...  Using subject retired metrics the error is fairly small on the current set (43 vs 2426 subjects), but some past subject sets are were very small (<100 subjects) and the error was large.  There are also many more accumulated classifications for the practice sets than for the current active "real" data at any time, so classification count metrics are also skewed.

**Comment by mrniaboc** [comment]
This continues to baffle even the most experienced Zooniverse researchers. The latest victims are our very own Helen Spiers and Cliff Johnson!

**Comment by hrspiers** [comment]
I'd second Grant's thoughts here - the progress bar is counterintuitive (I'm sure there's some sort of counter pun there...)

**Comment by eatyourgreens** [comment]
The new project home page shows completion for each workflow, which I think fixes this.
![Screenshot of Nest Quest GO, showing progress bars for each of their active workflows.](https://user-images.githubusercontent.com/59547/66199214-0ed78900-e696-11e9-8410-3fbba14c83cf.png)

**Comment by PmasonFF** [comment]
And none of the competed classification metrics deal with Caesar customized retirement rules in the EDC calculation, nor deal with "training" workflows with extremely high retirement rules on small subject sets.

What is needed is the average classifications to retire a subject for the active data for each workflow.  With that number and the average classifications per day for that workflow, one can give good completion estimates that reflect the work done and remaining in a linear fashion.

To get that number, one must look at the SubjectWorkflowStatus  attribute classification count for all retired subjects divided by the number of retired subjects.  This number would reflect the actual rules in place, and the actual experience of the project.  
This number is very stable for a given subject set,  workflow, and the project rules; it does not change rapidly once a significant number of subjects have retired. Hence it would not have to be updated frequently.  Early on, when a subject set has been changed or is new, and there is no data,  the retirement limit would serve as a proxy.

**Comment by PmasonFF** [comment]
I have tabulated the retirement completeness and the current classification completeness based on the current retirement limit, total subjects and total classifications for all active **workflows** for all live launched projects.

As of today there were 34 active workflow where either the classification completeness or the retired subject completeness was 100%  Of these only 17 show **both** completeness as 100% indicating that the classification metric is wrong about 50% of the time.
There are 5 cases where the retired completeness is 100% (workflow is finished) but the classification completeness is <100%, and 12 cases where the classification completeness  is >=100% but the retirement completeness is showing <100% and the workflow is not finished.

In addition there are at least 11 more situations (out of 171workflows) where the mismatch between the two metrics is such that there **has** to be an issue with the classification metric - many more could be inaccurate to a lesser degree.

There are three common ways the classification metric for a workflow gets messed up - changes to the retirement limit during the processing of a subject set, changes to the contents of active subject sets, and the activity of early retirement rules (Caesar).

To get a good classification based metric we would need to tabulate the current number of classifications needed to complete all unretired subjects vs those required to complete all subjects given the current subject list, current retirement limit and a correction for early retirement for the current subject mix.   This would require working at the subject level and tallying up classifications needed to complete each subject with some sort of allowance for Caesar based on its affect on the retirement of subjects up to that point.

This would be very resource costly and slowwww!  I am wondering if we can use the statistics of the underlying process to estimate a completeness metric based on modeling the process.

**Comment by yuenmichelle1** [comment]
Currently looking into this. (Excuse my rookie-ness and possible rant) I used https://www.zooniverse.org/projects/achmorrison/steelpan-vibrations/ to help me visualize and understand what the issue is. So from my understanding, the stats on the project's stats page is inconsistent with the completeness stat on the project's home page. (Due to different calculations for each) 

Again, using https://www.zooniverse.org/projects/achmorrison/steelpan-vibrations/ as an example. 
On the project's `/stats` page, completeness in this scenario is 
`= classifications / (subjects * retirement limit)` (at this point in time its  83678/(13119 * 15)=83678/196785 = approx 43%)

But on the project's home page, completeness is calculated as the average of `retired subjects / all subjects` for all  workflows.  In this case since there is 1 workflow, it would just be `retired / all`. (at this point in time, thats 3528/15119 = approx 26%) which looks less complete than 43% on `/stats` page.

The one issue I have here is that by the `/stats` pages calculations (again using numbers of steelpan vibrations at this point in time as an example), we have 83678 calculations, if we divide that by the retirement limit of 15, we technically should have approx 5578 retired subjects instead of 3528 retired subjects. 

So the question is, is  `classifications /  (classifications needed to retire a subject)` true "completeness" or is `retired / all subjects` in a workflow true "completeness". I dont know. 


One thing, i was thinking we could do (which might be what this issue is asking for?) is have a separate calculation similar to that of the `/stats` page and maybe have the project team decide which they would like to use as a calculation for completeness for their project's home page with the 'default' being the current way we are calculating on home page and adding maybe a disclaimer that numbers may be different due to multiple reasons (unretirement, early retirement, updating retirement limit, etc). (Or maybe it's enough to have a disclaimer that numbers may be different?)

**Comment by PmasonFF** [comment]
Yuenmichelle1,
The situation is much more complicated than your comment above:

1)	subjects are selected for classification pseudo randomly, so the retirement of a subject depends on chance – when that subject finally get selected enough times. Certainly total classification/retirement limit does not equal retired subjects until very near the end of the process.  The pseudo random process can be modeled, and surprisingly the relation between classification completeness and retirement completeness throughout the process depends only on the retirement limit – (though the smaller the number of subjects in a subject set, the more “noise” there is)  I generate a linearized completeness from the retirement completeness in my daily dashboard reports https://github.com/PmasonFF/Project_Health_Dashboard. by modeling the selection process for different retirement limits.

2)	How to state a project completeness is difficult – many projects have multiple workflows, some of which have no or very high retirement limits meant for training – including those into any completeness metric distorts the picture – for a project we really need a measure of how much work ( ie classification count) is needed to complete the research workflow(s) / the total work (ie classification count to retire all research subjects) with training workflows exempted.
3)	Caesar and early retirement make this exercise very difficult.  Early retirement rules eliminate an unknown set of subjects early in the process. This results in retired subject counts being higher than “normal” for the same classification count with normal retirement.  Because the percentage of subjects that will retire early is unknown, the process can not even be modeled accurately.

4)	The current project completeness is the average of the retirement completeness over all active workflows, and does not account for one or more workflows having larger active subject sets and hence require more work. 

What could be done to define a more useful project completeness:

First determine which workflows are to be included in the completeness metric in the project builder “show stats”, including only those workflows selected for visibility on the detailed stats page.

Then calculate the number of classifications required to complete all remaining subjects that are not retired assuming the retirement limit is required in each case.  Think of this as the work remaining for a workflow.  Note this in an expensive calculation and must be done by querying every unretired subject in the active subject sets to see how many classifications have been done and how many remain.

Finally a calculation of the total classifications done plus the work remaining gives a total work for the workflow/active subject sets.

Completeness for the workflow is (total work - work remaining)/total work. .  As the workflow progresses this metric self-corrects for the early retirement issues. (ie early retirement knocks out some subjects, those remaining are those where the retirement limit is a more accurate measure of the work remaining for that subject, and  the total work calculation correctly reflects the effect of early retirement on those subjects already retired.) 

Completeness for the project would be the (sum of total work for all included workflows – sum of the work remaining for all included workflows)/ sum of total work for all included workflows.

**Comment by NomadPurple** [comment]
Adding to the discussion from my experience on calculating an estimate of remaining classifications for Features where I need estimated days to completion.
Pros of classification completion.
1. Motivating for volunteers/classifiers as progress bar will start moving much earlier eg. on a flat retirement limit of 15, subject completion doesn't start until around 45-50% of classification completion.
2. It MIGHT be easier to understand - caveat: the change to a straight % of Total project subjects completed has made progress bars cleared than before, given that around 60% of projects have multiple workflows during their lifespan.

Cons
1. The complexity of calculation - as explained by Pmason above, it's difficult.
I use the basic calculation for each workflow of total subj x ret limit - classifications done
but the problems encountered are
1. Removal of Practice workflows
2. Workflows with no retirement limit (currently 8% of projects) so I use an estimate or if genuinely no retirement I have to exclude.
3. Caesar, Group transcription and other retirement rules - I use estimates of average retirement limit from previous workflows/seasons/experience. Retirement limits are often non-linear over time within a workflow, with lower retirement classifications at the start for human images or blank images/pages to final subjects being more difficult requiring more classifications.
4. Workflows where the retirement limit changed at some point in the past - I set max/min points to detect these workflows and then use the rate of subject ret/classifications to estimate classifications on remaining subjects. It's a very rough calculation which isn't reliable until subjects start retiring but it's the best I can do with reasonable efficiency.
5.  Workflows which in the past have received a great many classification on retired subjects, usually when in Paused status, and then have new subjects added without removing the old batched (Steelpan vibrations is an example of this). This gets worse over time when projects are never 'zero-ed" and my solution is the same as for 4. above. The problem with a progress bar on these is that it would look like far more progress has been made than true and at some point the bar would track down/back as it adjusted for the error. Not a good for motivation or simplicity. (If I can suggest to projects to zero for new data sets I do, but with limited success)

I worked through today's active projects and conditions 3 ,4 and 5 are true for at least 37% of projects so cannot be ignored.
